### ISSUE TITLE
linked params do not follow parameter limits: was segfault with linked parameters + atables

### ISSUE THREAD

**Issue by JohannesBuchner** [sherpa/sherpa#742]
I seem to be getting segfaults when setting up a atable (with load_table_model), and linking one of its parameters to another parameter of another component. When that parameter exceeds the limit of the table parameters, it tries to look up data that is out of bounds, leading to a segfault.
The correct behavior should prevent the table from exceeding its parameter limits.

**Comment by dtnguyen2** [comment]
Can you make script and data available, thanks in advance

**Comment by JohannesBuchner** [comment]
This script should reproduce the issue with any X-ray spectrum file.
```
from sherpa.astro.ui import *
import os
import numpy

filename = 'g20_a.pha'
load_pha(filename)

# from https://zenodo.org/record/2245189
load_table_model('blob', '/home/user/bin/xrayspectralmodels/clumpybox3.fits')
set_model(blob + xspowerlaw.p * xstbabs.abso)
blob.norm = 1e-10

from sherpa.models.parameter import Parameter
# make a parameter with wide limits
srcnh = Parameter('src', 'nh', 22, 19, 27, 19, 27)
# link to one component
xstbabs.abso.nh = 10**(srcnh - 22)
print(get_model())

print("test with single link ...")
print(calc_stat())
print("test with double link ...")
# link to a link, but blob.nh does not support that wide of a range
blob.nh.link = xstbabs.abso.nh
print(calc_stat()) # --segfault occurs here

```

**Comment by dtnguyen2** [comment]
I got tired of waiting to download your large _clumpybox3.fits_ file and substitute a fits file that I have around but the results, apart from the output of the ```calc_stat```, should not affect the issue you are asking.  Anyway, not sure how your script manages to get the following line to work:
```
blob.norm = 1e-10
```
because sherpa's models usually use ```ampl``` instead of ```norm```;  Moreover, notice also that blob does not have parameter  called ```nh``` either.  I double checked sherpa's class definition of the TableModel:
```
class TableModel(ArithmeticModel):
    def __init__(self, name='tablemodel'):
        # these attributes should remain somewhat private
        # as not to conflict with user defined parameter names
        self.__x = None
        self.__y = None
        self.__filtered_y = None
        self.filename = None
        self.method = linear_interp  # interpolation method
        self.ampl = Parameter(name, 'ampl', 1)
        ArithmeticModel.__init__(self, name, (self.ampl,))
...
(Pdb) blob
<TableModel model instance 'tablemodel.blob'>
(Pdb) print(blob)
tablemodel.blob
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   blob.ampl    thawed        1e-10 -3.40282e+38  3.40282e+38           
```

As a result, I get the following output (ie not a segfault):

62862.88126117057
test with double link ...
Traceback (most recent call last):
  File "tst.py", line 44, in <module>
    blob.nh.link = xstbabs.abso.nh
  File "/export/sherpa/sherpa/models/model.py", line 190, in __getattr__
    NoNewAttributesAfterInit.__getattribute__(self, name)
AttributeError: 'TableModel' object has no attribute 'nh'

**Comment by JohannesBuchner** [comment]
I am using sherpa 4.12.0 installed via conda if that helps.

**Comment by JohannesBuchner** [comment]
Is your fits file a xspec atable?

**Comment by JohannesBuchner** [comment]
Apparently the function has been renamed load_xstable_model in recent releases.

**Comment by dtnguyen2** [comment]
I can reproduce the segfault, will look into this

**Comment by dtnguyen2** [comment]
The [url](https://zenodo.org/record/2245189) only contains the file _clumpybox3.fits_ and the file _g20_a.pha_ cannot be found at the aforementioned url.  So I improvised by using the file _data.txt_ which is available in the sherpa-test-data directory; However, I don't think by using a _data.txt_ file will change the issue at hand (aside from the printed stat value) because I can reproduce the segfault using the original script.

The default values for the ```nh``` parameter of the XSTableModel are:
```
val         = 100.0
min         = 28.200000762939453
max         = 10000.0
units       = 
frozen      = False
link        = None
default_val = 100.0
default_min = 28.200000762939453
default_max = 10000.0
``` 
Notice that the default value of ```nh``` is to range within ```[28.2,  10000.0]```.  I modified the original script to make it to run within xspec's default values:
```
(sherpa) [dtn@devel12 dtn]$ diff bucher.py withinlimits.py
24c24
< srcnh = Parameter('src', 'nh', 22, 19, 27, 19, 27)
---
> srcnh = Parameter('src', 'nh', 100, 28.3, 10000, 28.3, 10000)
34c34
< xstbabs.abso.nh = 10**(srcnh - 22)
---
> xstbabs.abso.nh = (srcnh - 22)
```

As one can see (aside from a bunch of useless info deleted), the script runs to completion, ie no segfault:

```
(sherpa) [dtn@devel12 dtn]$ python withinlimits.py 
...
test with double link ...
(xstablemodel.blob + (xspowerlaw.p * xstbabs.abso))
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   blob.nh      linked           78            expr: abso.nH           
....         
   abso.nH      linked           78      expr: (src.nh - 22) 10^22 atoms / cm^2
val         = 78.0
min         = 28.200000762939453
max         = 10000.0
units       = 
frozen      = True
link        = abso.nH
default_val = 78.0
default_min = 28.200000762939453
default_max = 10000.0
tbvabs Version 2.3
...
Note that this routine ignores the current cross section setting
as it always HAS to use the Verner cross sections as a baseline.
65357.005427474425
```
The _linking one of its parameters to another parameter of another component_ is not the source of the segfault. In the next script, nolink_outside_limits.py, the two sherpa commands to link the ```nh``` parameter are commented out; Moreover, the ```blob.nh``` parameter is set to ```22``` (outside of the default limits of ```[28.2,  10000.0]```).

```
(sherpa) [dtn@devel12 dtn]$ diff bucher.py nolink_outside_limits.py 
34c34
< xstbabs.abso.nh = 10**(srcnh - 22)
---
> # xstbabs.abso.nh = 10**(srcnh - 22)
40c40
< blob.nh.link = xstbabs.abso.nh
---
> # blob.nh.link = xstbabs.abso.nh
41a42,44
> blob.nh._hard_min = 17
> blob.nh.min = 17
> blob.nh.val = 22
45d47
< 
```
One can override the xspec's default limits, but it seems like the default limits of ```[28.2,  10000.0]``` are set for a good reason:

```
(sherpa) [dtn@devel12 dtn]$ python nolink_outside_limits.py
...
test with double link ...
(xstablemodel.blob + (xspowerlaw.p * xstbabs.abso))
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   blob.nh      thawed           22           17        10000           
...
   abso.nH      thawed            1            0       100000 10^22 atoms / cm^2
val         = 22.0
min         = 17.0
max         = 10000.0
units       = 
frozen      = False
link        = None
default_val = 22.0
default_min = 28.200000762939453
default_max = 10000.0
Segmentation fault
```

In summary, the _linking one of its parameters to another parameter of another component_ is not the source of the segfault.  While one can override the xspec's ```xstablemodel nh``` default limits, unfortunately the source of the segfault is due to the fact that the parameter  ```xstablemodel nh``` is used outside the default limits.

**Comment by JohannesBuchner** [comment]
A segfault is not a good behaviour in out-of-limit cases; it would be better if xstablemodel, instead of trying to access invalid memory, raised an error.

**Comment by dtnguyen2** [comment]
So I added the following two print statements as follows:
```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/models/model.py 
diff --git a/sherpa/models/model.py b/sherpa/models/model.py
index 5e4c0bf4..8709e1bd 100644
--- a/sherpa/models/model.py
+++ b/sherpa/models/model.py
@@ -64,9 +64,9 @@ def modelCacher1d(func):
             digest = hashlib.sha256(token).digest()
             if digest in cache:
                 return cache[digest].copy()
-
+        print('will call func', func)
         vals = func(cls, pars, xlo, *args, **kwargs)
-
+        print('returned from func', func)
         if use_caching:
             # remove first item in queue and remove from cache
             key = queue.pop(0)
```
so when I run the original script, I get the following output:
```
...
default_min = 28.200000762939453
default_max = 10000.0
will call func <function XSTableModel.calc at 0x7fa763beea60>
Segmentation fault
```
Notice that the print statement:
```
        print('returned from func', func)
```
was not executed. So the segfault is within the xspec code.

**Comment by JohannesBuchner** [comment]
XSTableModel is not xspec code, it is a sherpa interface to xspec tables. 

It should do a within-bounds check in calc here: https://sherpa.readthedocs.io/en/latest/_modules/sherpa/astro/xspec.html#XSTableModel

**Comment by dtnguyen2** [comment]
```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/astro/xspec/__init__.py
diff --git a/sherpa/astro/xspec/__init__.py b/sherpa/astro/xspec/__init__.py
index 689f6fe6..07519a6e 100644
--- a/sherpa/astro/xspec/__init__.py
+++ b/sherpa/astro/xspec/__init__.py
@@ -953,9 +953,12 @@ class XSTableModel(XSModel):
         #
         if hasattr(_xspec, 'tabint'):
             tabtype = 'add' if self.addmodel else 'mul'
-            return _xspec.tabint(p,
-                                 filename=self.filename, tabtype=tabtype,
-                                 *args, **kwargs)
+            print('hello world')
+            tmp = _xspec.tabint(p,
+                                filename=self.filename, tabtype=tabtype,
+                                *args, **kwargs)
+            print('goodbye cruel world')
+            return tmp
 
         if self.addmodel:
             func = _xspec.xsatbl
```
```
(sherpa) [dtn@devel12 dtn]$ python bucher.py 
...
default_min = 28.200000762939453
default_max = 10000.0
hello world
Segmentation fault
```
In the file _sherpa/astro/xspec/src/_xspec.cc_ one finds:

```
// XSPEC table models; in XSPEC 12.10.1 these have been consolidated
// into the tabint routine.
//
#ifndef XSPEC_12_10_1
void xsatbl(float* ear, int ne, float* param, const char* filenm, int ifl, 
	    float* photar, float* photer);
void xsmtbl(float* ear, int ne, float* param, const char* filenm, int ifl, 
	    float* photar, float* photer);
```
In the file _sherpa/include/sherpa/astro/xspec_extension.hh_
```
#ifdef XSPEC_12_10_1
void tabint(float* ear, int ne, float* param, int npar, const char* filenm, int ifl,
             const char* tabtyp, float* photar, float* photer);
#endif
```

```
#ifdef XSPEC_12_10_1
#define XSPECTABLEMODEL        \
		{ (char*)"tabint", \
	(PyCFunction)((PyCFunctionWithKeywords)sherpa::astro::xspec::xspectablemodel), \
	METH_VARARGS|METH_KEYWORDS, \
	NULL }
```

**Comment by JohannesBuchner** [comment]
Yes, therefore it would be good to check if the parameters in this function are valid inputs for the table interpolation.

I would suggest inserting into calc() something like:

`assert all((param.min <= value <= param.max for value, param in zip(self.pars, p))), "model parameters outside bounds"`

**Comment by DougBurke** [comment]
I've changed the title because the underlying issue is that linked parameters do not enforce their own min/max value, but that of the linked parameter. So ideally we'd fix it generally rather than just in the XSPEC table model code (as #743 does).

To see the issue, I create power-law and constant models - neither of which are from XSPEC - and set the power-law slope (which is restricted -10 to 10) and link it to the constant model (which has a lmiit of -3e38 to +3e38 or so). The -10/10 range is then not enforced, but you can see the ampl range (>= 0) is still enforced:

```
In [1]: from sherpa.models.basic import Const1D, PowLaw1D                                                                                                                                          

In [2]: m1 = PowLaw1D()                                                                                                                                                                            

In [3]: m2 = Const1D()                                                                                                                                                                             

In [4]: print(m1)                                                                                                                                                                                  
powlaw1d
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   powlaw1d.gamma thawed            1          -10           10           
   powlaw1d.ref frozen            1 -3.40282e+38  3.40282e+38           
   powlaw1d.ampl thawed            1            0  3.40282e+38           

In [5]: print(m2)                                                                                                                                                                                  
const1d
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   const1d.c0   thawed            1 -3.40282e+38  3.40282e+38           

In [6]: m1.gamma = m2.c0                                                                                                                                                                           

In [7]: print(m1)                                                                                                                                                                                  
powlaw1d
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   powlaw1d.gamma linked            1         expr: const1d.c0           
   powlaw1d.ref frozen            1 -3.40282e+38  3.40282e+38           
   powlaw1d.ampl thawed            1            0  3.40282e+38           

In [8]: m2.c0 = 2                                                                                                                                                                                  

In [9]: print(m1)                                                                                                                                                                                  
powlaw1d
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   powlaw1d.gamma linked            2         expr: const1d.c0           
   powlaw1d.ref frozen            1 -3.40282e+38  3.40282e+38           
   powlaw1d.ampl thawed            1            0  3.40282e+38           

In [10]: m2.c0 = 20                                                                                                                                                                                

In [11]: print(m1)                                                                                                                                                                                 
powlaw1d
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   powlaw1d.gamma linked           20         expr: const1d.c0           
   powlaw1d.ref frozen            1 -3.40282e+38  3.40282e+38           
   powlaw1d.ampl thawed            1            0  3.40282e+38           

In [12]: m1.ampl = -1                                                                                                                                                                              
---------------------------------------------------------------------------
ParameterErr                              Traceback (most recent call last)
<ipython-input-12-2857c1b3c141> in <module>
----> 1 m1.ampl = -1

/lagado.real/sherpa/sherpa-py38/sherpa/models/model.py in __setattr__(self, name, val)
    195             # When setting an attribute that is a Parameter, set the parameter's
    196             # value instead.
--> 197             par.val = val
    198         else:
    199             NoNewAttributesAfterInit.__setattr__(self, name, val)

/lagado.real/sherpa/sherpa-py38/sherpa/utils/__init__.py in __setattr__(self, name, val)
    157                                      (type(self).__name__, name))
    158 
--> 159         object.__setattr__(self, name, val)
    160 
    161 

/lagado.real/sherpa/sherpa-py38/sherpa/models/parameter.py in _set_val(self, val)
    165             val = SherpaFloat(val)
    166             if val < self.min:
--> 167                 raise ParameterErr('edge', self.fullname, 'minimum', self.min)
    168             if val > self.max:
    169                 raise ParameterErr('edge', self.fullname, 'maximum', self.max)

ParameterErr: parameter powlaw1d.ampl has a minimum of 0

```

**Comment by DougBurke** [comment]
Continuing my comment: if you "break" the link then the -10/10 range is enforced.

```
In [13]: m1.gamma = 3                                                                                                                                                                              

In [14]: m1.gamma = 12                                                                                                                                                                             
---------------------------------------------------------------------------
ParameterErr                              Traceback (most recent call last)
<ipython-input-14-161f13cdb9ca> in <module>
----> 1 m1.gamma = 12

/lagado.real/sherpa/sherpa-py38/sherpa/models/model.py in __setattr__(self, name, val)
    195             # When setting an attribute that is a Parameter, set the parameter's
    196             # value instead.
--> 197             par.val = val
    198         else:
    199             NoNewAttributesAfterInit.__setattr__(self, name, val)

/lagado.real/sherpa/sherpa-py38/sherpa/utils/__init__.py in __setattr__(self, name, val)
    157                                      (type(self).__name__, name))
    158 
--> 159         object.__setattr__(self, name, val)
    160 
    161 

/lagado.real/sherpa/sherpa-py38/sherpa/models/parameter.py in _set_val(self, val)
    167                 raise ParameterErr('edge', self.fullname, 'minimum', self.min)
    168             if val > self.max:
--> 169                 raise ParameterErr('edge', self.fullname, 'maximum', self.max)
    170 
    171             self._val = val

ParameterErr: parameter powlaw1d.gamma has a maximum of 10

```

This may be tricky to enforce/restrict (since a parameter doesn't know what links to it, so can't enforce anything at that point). Maybe we can add some logic to the `_get_val` method of `sherpa.models.parameter.Parameter`?

**Comment by DougBurke** [comment]
So, I have a "general" fix for this issue in #745 

See the discussion on #743 (which is a lower-level fix specific to XSPEC table models).
